Add documentation about .gdextension file (#9383)
godotengine/godot-docs: tutorials/scripting/gdextension/gdextension_file.rst

.. _doc_gdextension_file_sections:

The .gdextension file
=====================

Introduction
------------

The ``.gdextension`` file in your project contains the instructions for how to load
the GDExtension. The instructions are separated into specific sections. This page
should give you a quick overview of the different options available to you. For an introduction
how to get started with GDExtensions take a look at the :ref:`GDExtension C++ Example <doc_gdextension_cpp_example>`.

Configuration section
---------------------

+-------------------------------+------------------------------------------------------------------------------------------------------+
| Instruction                   | Description                                                                                          |
+===============================+======================================================================================================+

> | **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |

Reviewer: ```suggestion
| **entry_symbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |
```

Reviewer 2: ```suggestion
| **entry_symbol**              | Name of the entry function for initializing the GDExtension. This function should be defined in      |
```
Rather

|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
+-------------------------------+------------------------------------------------------------------------------------------------------+

> | **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |

Reviewer: ```suggestion
| **compatibility_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
```

|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
+-------------------------------+------------------------------------------------------------------------------------------------------+

> | **compatability_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |

Reviewer: ```suggestion
| **compatibility_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |
```

+-------------------------------+------------------------------------------------------------------------------------------------------+
| **android_aar_plugin**        | The GDExtension is part of a :ref:`v2 Android plugin <doc_android_plugin>`. During export this flag  |
|                               | will indicate to the editor that the GDExtension native shared libraries are exported by the Android |
|                               | plugin AAR binaries.                                                                                 |
+-------------------------------+------------------------------------------------------------------------------------------------------+

Libraries section
-----------------

In this section you can set the paths to the compiled binaries of your GDExtension libraries.
By setting feature flags you can filter which version should be loaded and exported with your
game depending on which feature flags are active. Every feature flag must match to Godot's
feature flags. For instance ``macos.debug`` means that it will be loaded if Godot has both the

> ``macos`` and ``debug`` flag active. Each line of the section is evaluated from top to bottom.

Reviewer 2: Does this not support custom feature flags on export? This and below implies only the official ones are accepted

+-------------------------------+------------------------------------------------------------------------------------------------------+
| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **compatability_maximum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.3+**                   |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |

> |                               | depend on features from newer versions of Godot.                                                     |

Reviewer 3: Duplicate `compatability_minimum`

| Flag                          | Description                                                                                          |
+===============================+======================================================================================================+
| **double**                    | double-precision build                                                                               |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **single**                    | single-precision build                                                                               |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **64**                        | 64-bit build (any architecture)                                                                      |
+-------------------------------+------------------------------------------------------------------------------------------------------+

> | **32**                        | 32-bit build (any architecture)                                                                      |

Reviewer 3: Given that, for example, x86_64 and arm64 are not binary-compatible, I don't think it's worth mentioning/encouraging `64` or `32` as options. I think we should link to Godot's feature tag documentation for a complete list, and on this documentation page only list the ones we encourage using for GDExtension. Same with `x86` and `arm`, those should probably not be used with GDExtension, instead `arm64` etc should be used.

Also, Godot doesn't support 64-bit WebAssembly yet.

Author: I removed the feature tags that should not be advertised as prominently as suggested

.. _doc_gdextension_file_sections:

The .gdextension file
=====================

Introduction
------------

The ``.gdextension`` file in your project contains the instructions for how to load
the GDExtension. The instructions are separated into specific sections. This page
should give you a quick overview of the different options available to you. For an introduction
how to get started with GDExtensions take a look at the :ref:`GDExtension C++ Example <doc_gdextension_cpp_example>`.

Configuration section
---------------------

+-------------------------------+------------------------------------------------------------------------------------------------------+

> | Instruction                   | Description                                                                                          |

Reviewer 4: I'm not sure about calling these "Instruction"s. Maybe "Property" instead?

```suggestion
| Property                  | Description                                                                                          |
```

+===============================+======================================================================================================+
| **entry_symbol**              | Name of the entry function for initializing the GDExtension. This function should be defined in      |
|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **compatibility_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **compatibility_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |

> |                               | **Only supported for Godot 4.3+**                                                                    |

Reviewer 3: "or later" might also be good to add.

+-------------------------------+------------------------------------------------------------------------------------------------------+
| **reloadable**                | Reloads the extension upon recompilation. Reloading is not supported for every exposed class.        |

> |                               | This flag should be mainly used for developing or debugging an extension.                            |

Reviewer 4: What is meant by "Reloading is not supported for every exposed class"? It should work for any class, although, some can have additional considerations, for example, editor plugins.

It may be worth pointing out that not all GDExtension bindings support it? We could mention that godot-cpp for Godot 4.2+ does - I'm not sure it makes sense trying to list other bindings, because that list will always be growing

Author: I wasn't sure if it is already supported for every class especially the editor plugins classes due to how they are handled with reloading. If are you certain then I will change it. I would only list the official godot-cpp support as we other language bindings can say themselves if they support reload support or not

Author: Updated now. Let me know what you think

Reviewer 4: Looks good to me!

+-------------------------------+------------------------------------------------------------------------------------------------------+

> | **android_aar_plugin**        | The GDExtension is part of a :ref:`v2 Android plugin <doc_android_plugin>`. During export this flag  |

Reviewer 4: It's not clear from this description that this is a boolean property. Perhaps we should be listing the type for each property?

Reviewer 4: Thanks!

|                               | will indicate to the editor that the GDExtension native shared libraries are exported by the Android |
|                               | plugin AAR binaries.                                                                                 |
+-------------------------------+------------------------------------------------------------------------------------------------------+

Libraries section
-----------------

In this section you can set the paths to the compiled binaries of your GDExtension libraries.

> By setting feature flags you can filter which version should be loaded and exported with your

Reviewer 4: ```suggestion
By specifying feature flags you can filter which version should be loaded and exported with your
```

game depending on which feature flags are active. Every feature flag must match to Godot's

> feature flags or your custom export flags to run in an exported game. For instance ``macos.debug``

Reviewer 4: ```suggestion
feature flags or your custom export flags to be loaded in an exported game. For instance ``macos.debug``
```

line of the section is evaluated from top to bottom.

Here is an example of what that can look like:

.. code-block:: none

    [libraries]

    macos.debug = "res://bin/libgdexample.macos.template_debug.framework"
    macos.release = "res://bin/libgdexample.macos.template_release.framework"
    windows.debug.x86_32 = "res://bin/libgdexample.windows.template_debug.x86_32.dll"
    windows.release.x86_32 = "res://bin/libgdexample.windows.template_release.x86_32.dll"
    windows.debug.x86_64 = "res://bin/libgdexample.windows.template_debug.x86_64.dll"
    windows.release.x86_64 = "res://bin/libgdexample.windows.template_release.x86_64.dll"
    linux.debug.x86_64 = "res://bin/libgdexample.linux.template_debug.x86_64.so"
    linux.release.x86_64 = "res://bin/libgdexample.linux.template_release.x86_64.so"
    linux.debug.arm64 = "res://bin/libgdexample.linux.template_debug.arm64.so"
    linux.release.arm64 = "res://bin/libgdexample.linux.template_release.arm64.so"
    linux.debug.rv64 = "res://bin/libgdexample.linux.template_debug.rv64.so"
    linux.release.rv64 = "res://bin/libgdexample.linux.template_release.rv64.so"

> Here are lists of available built-in options (for more look at the :ref:`feature tags <doc_feature_tags>`):

Reviewer 4: ```suggestion
Here are lists of some of the available built-in options (for more look at the :ref:`feature tags <doc_feature_tags>`):
```

.. _doc_gdextension_file_sections:

The .gdextension file
=====================

Introduction
------------

> The ``.gdextension`` file in your project contains the instructions for the editor how to parse
> the GDExtensions in projects. The instructions are separated into specific sections. This page

Reviewer 4: ```suggestion
The ``.gdextension`` file in your project contains the instructions for how to load
the GDExtension. The instructions are separated into specific sections. This page
```

should give you a quick overview of the different options available to you. For an introduction
how to get started with GDExtensions take a look at the :ref:`GDExtension C++ Example <doc_gdextension_cpp_example>`.

Configuration section
---------------------

+-------------------------------+------------------------------------------------------------------------------------------------------+
| Instruction                   | Description                                                                                          |
+===============================+======================================================================================================+

> | **entry_smbol**               | Name of the entry function of the loaded GDExtension. This function should be defined in             |

Reviewer 4: ```suggestion
| **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in             |
```

|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
+-------------------------------+------------------------------------------------------------------------------------------------------+

> | **compatability_minimum**     | Minimum compatible version. This prevents older editor binaries from loading the extension.          |

Reviewer 4: ```suggestion
| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that depend on features from newer versions of Godot.          |
```

+-------------------------------+------------------------------------------------------------------------------------------------------+

Libraries section
-----------------

> In this section you can set the paths to your compiled binaries of your GDExtension libraries.

Reviewer 4: ```suggestion
In this section you can set the paths to the compiled binaries of your GDExtension libraries.
```

By setting feature flags you can filter which version should be loaded and exported with your

> game depending on which feature flags are active.

Reviewer 4: I think we should add something about how all feature flags on each line must match (so, `macos.debug` means that it will be loaded if Godot has both "macos" and "debug" feature flags), and that each line is evaluated in order from top-to-bottom.

Debugging symbols
^^^^^^^^^^^^^^^^^

+-------------------------------+------------------------------------------------------------------------------------------------------+
| Flag                          | Description                                                                                          |
+===============================+======================================================================================================+
| **debug**                     | Build with debug symbols                                                                             |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **release**                   | Optimized build without debug symbols                                                                |
+-------------------------------+------------------------------------------------------------------------------------------------------+

Architecture
^^^^^^^^^^^^

+-------------------------------+------------------------------------------------------------------------------------------------------+

> | Flag                          | Description                                                                                          |
> +===============================+======================================================================================================+
> | **x86_64**                    | 64-bit Windows                                                                                       |
> +-------------------------------+------------------------------------------------------------------------------------------------------+
> | **x86_32**                    | 32-bit Windows                                                                                       |

Reviewer 3: The flags are architecture-specific, not Windows-specific. x86_64 means 64-bit x86, not 64-bit Windows.

```suggestion
| Flag                          | Description                                                                                          |
+===============================+======================================================================================================+
| **x86_64**                    | 64-bit x86                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **x86_32**                    | 32-bit x86                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **arm32**                     | 32-bit Arm                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **arm64**                     | 64-bit Arm                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **rv64**                      | 64-bit RISC-V                                                                                        |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **ppc32**                     | 32-bit PowerPC                                                                                       |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **ppc64**                     | 64-bit PowerPC                                                                                       |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **wasm32**                    | 32-bit WebAssembly                                                                                   |
```

Author: Noticed already. Thank you!
